integration-testing/src/test/kotlin/ListAllCoroutineThrowableSubclassesTest.kt

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2016-2022 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines
+
+import com.google.common.reflect.*
+import kotlinx.coroutines.*
+import org.junit.Test
+import kotlin.test.*
+
+class ListAllCoroutineThrowableSubclassesTest {
+
+    /*
+     * These are all known throwables in kotlinx.coroutines.
+     * If you have added one, this test will fail to make
+     * you ensure your exception type is java.io.Serializable.
+     *
+     * We do not have means to check it automatically, so checks are delegated to humans.
+     *
+     * See #3328 for serialization rationale.
+     */
+    private val knownThrowables = setOf(
+        "kotlinx.coroutines.TimeoutCancellationException",
+        "kotlinx.coroutines.JobCancellationException",
+        "kotlinx.coroutines.internal.UndeliveredElementException",
+        "kotlinx.coroutines.CompletionHandlerException",
+        "kotlinx.coroutines.DiagnosticCoroutineContextException",
+        "kotlinx.coroutines.CoroutinesInternalError",
+        "kotlinx.coroutines.channels.ClosedSendChannelException",
+        "kotlinx.coroutines.channels.ClosedReceiveChannelException",
+        "kotlinx.coroutines.flow.internal.ChildCancelledException",
+        "kotlinx.coroutines.flow.internal.AbortFlowException",
+        )
+
+    @Test
+    fun testThrowableSubclassesAreSerializable() {
+        var throwables = 0
+        val classes = ClassPath.from(this.javaClass.classLoader)
+            .getTopLevelClassesRecursive("kotlinx.coroutines");
+        classes.forEach {
+            if (Throwable::class.java.isAssignableFrom(it.load())) {
+                ++throwables
+                // println(""""$it",""")
+                assertTrue(knownThrowables.contains(it.toString()))
+            }
+        }
+
+        assertEquals(knownThrowables.size, throwables)
```

Comment:
I wanted to propose this code, which I find much clearer:
```suggestion
        val classes = ClassPath.from(this.javaClass.classLoader)
            .getTopLevelClassesRecursive("kotlinx.coroutines");
        val throwables = classes.filter { Throwable::class.java.isAssignableFrom(it.load()) }
            .map { it.toString() }.toSet()
        assertEquals(knownThrowables, throwables)
```
However, it turns out, the errors arising from set comparisons are displayed very poorly.
```
expected:<[kotlinx.coroutines.TimeoutCancellationException, kotlinx.coroutines.JobCancellationException, kotlinx.coroutines.internal.UndeliveredElementException, kotlinx.coroutines.CompletionHandlerException, kotlinx.coroutines.DiagosticCoroutineContextException, kotlinx.coroutines.CoroutinesInternalError, kotlinx.coroutines.channels.ClosedSendChannelException, kotlinx.coroutines.channels.ClosedReceiveChannelException, kotlinx.coroutines.flow.internal.ChildCancelledException, kotlinx.coroutines.flow.internal.AbortFlowException]> but was:<[kotlinx.coroutines.DiagnosticCoroutineContextException, kotlinx.coroutines.JobCancellationException, kotlinx.coroutines.internal.UndeliveredElementException, kotlinx.coroutines.TimeoutCancellationException, kotlinx.coroutines.CompletionHandlerException, kotlinx.coroutines.CoroutinesInternalError, kotlinx.coroutines.flow.internal.ChildCancelledException, kotlinx.coroutines.flow.internal.AbortFlowException, kotlinx.coroutines.channels.ClosedReceiveChannelException, kotlinx.coroutines.channels.ClosedSendChannelException]>
Expected :[kotlinx.coroutines.TimeoutCancellationException, kotlinx.coroutines.JobCancellationException, kotlinx.coroutines.internal.UndeliveredElementException, kotlinx.coroutines.CompletionHandlerException, kotlinx.coroutines.DiagosticCoroutineContextExcepti ...

Actual   :[kotlinx.coroutines.DiagnosticCoroutineContextException, kotlinx.coroutines.JobCancellationException, kotlinx.coroutines.internal.UndeliveredElementException, kotlinx.coroutines.TimeoutCancellationException, kotlinx.coroutines.CompletionHandlerExcept ...
```
Maybe a good reason to file an issue somewhere?

Reply:
Yes, that's IDEA subsystem in YT.

I do not have strong opinion about how this test should be written, feel free to commit your suggestion if you believe it's better

Reply:
I don't want to commit that, because, as you can see, the errors are just unusable in this form. The current version, while just being an approximation of what it wants to achieve—set comparisons—at least works.